_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+
+First though, let me again set the context. In physics, an all-important concept is that of symmetry[^1], a symmetry being present whenever some quantity is being conserved. But we don't even need to look to science. Examples arise in daily life, and -- otherwise why write about it - in the tasks we apply deep learning to. In daily life: Think about speech -- me stating "it is cold", for example. Formally, or denotation-wise, the sentence will have the same meaning now as in five hours. (Connotations, on the other hand, can and will probably be different!). This is a form of translation symmetry, translation in time. In deep learning: Take image classification. For the usual convolutional neural network, a cat in the center of the image is just that, a cat; a cat on the bottom is, too. But one sleeping, comfortably curled like a half-moon "open to the right", will not be "the same" as one in a mirrored position. Of course, we can train the network to treat both as equivalent by providing training images of cats in both positions, but that is not a scaleable approach. Instead, we'd like to make the network aware of these symmetries, so they are automatically preserved throughout the network architecture.
+
+[^1]: This is nicely explained in, for example, Jakob Schwichtenberg's [Physics from symmetry](http://physicsfromsymmetry.com/).
+
+## Purpose and scope of this post
+
+Here, I introduce [`escnn`](https://github.com/QUVA-Lab/escnn), a PyTorch extension that implements forms of group equivariance for CNNs operating on the plane or in (3d) space. The library is used in various, amply illustrated research papers; it is appropriately documented; and it comes with introductory notebooks both relating the math and exercising the code. Why, then, not just refer to the [first notebook](https://github.com/QUVA-Lab/escnn/blob/master/examples/introduction.ipynb), and immediately start using it for some experiment?
+
+In fact, this post should -- as quite a few texts I've written -- be regarded as an introduction to an introduction. To me, this topic seems anything but easy, for various reasons. Of course, there's the math. But as so often in machine learning, you don't need to go to great depths to be able to apply an algorithm correctly. So if not the math itself, what generates the difficulty? For me, it's two things.
```

Comment:
This is a bit hard to follow in terms of what you're setting up for the paragraphs that come after—that there are two things about "this topic" (not totally sure what that is, but I'm also proofing, so maybe that's my bad) that make it difficult, but it's not the math?

Reply:
Yes, exactly